### PR TITLE
add support for arbritary remote storages

### PR DESCRIPTION
**Pull request by kayibal**
Mainly to support gcsfs

### CODE REVIEW

File: sparsity/io.py [hunk 1]
```diff
@@ -11,6 +12,30 @@
...
+
+    open = open
+
+FILE_SYSTEMS = {
+    '': LocalFileSystem,
+    'file': LocalFileSystem
+}
+
+try:
+    import s3fs
```
Comment (michcio1234) [inline sparsity/io.py]: There's already `s3fs` import above. Wrapping this one in `try` doesn't seem to make sense.

File: sparsity/io.py [hunk 2]
```diff
@@ -28,38 +53,45 @@ def traildb_to_coo(db, fieldname):
...
         {'multiindex': True if isinstance(sf.index, pd.MultiIndex) else False}
     data['frame_index'] = sf.index.values
     data['frame_columns'] = sf.columns.values
     if not filename.endswith('.npz'):
         filename += '.npz'
-    if not filename.startswith('s3://'):
+
+    protocol = urlparse(filename).scheme or 'file'
+    if protocol == 'file':
         fp = open(filename, 'wb')
```
Comment (michcio1234) [inline sparsity/io.py]: Do we leave the file open here? O.o

File: sparsity/io.py [hunk 3]
```diff
@@ -28,38 +53,45 @@ def traildb_to_coo(db, fieldname):
...
+        buffer = BytesIO()
+        np.savez(buffer, **data)
+        buffer.seek(0)
+        _save_remote(buffer, filename, block_size, storage_options)
+
+
+def _save_remote(buffer, filename, block_size=None, storage_options=None):
+    if storage_options is None:
+        storage_options = {}
+    protocol = urlparse(filename).scheme or 'file'
```
Comment (michcio1234) [inline sparsity/io.py]: `scheme` shouldn't be `'file'` here. I think leaving this as an option is misleading.

File: sparsity/io.py [hunk 4]
```diff
@@ -28,38 +53,45 @@ def traildb_to_coo(db, fieldname):
...
-    open_f = open if not filename.startswith('s3://') \
-        else S3FileSystem().open
+def read_npz(filename, storage_options=None):
+    if storage_options is None:
+        storage_options = {}
+    protocol = urlparse(filename).scheme or 'file'
+    open_f = FILE_SYSTEMS[protocol](**storage_options).open
     fp = open_f(filename, 'rb')
 
     loader = np.load(fp)
```
Comment (michcio1234) [inline sparsity/io.py]: Same here - we're not closing the file.

File: sparsity/io.py (hunk 1 above)
Comment (kayibal) [reply sparsity/io.py]: true I'll remove the one above

File: sparsity/io.py (hunk 3 above)
Comment (kayibal) [reply sparsity/io.py]: that's correct

File: sparsity/io.py (hunk 4 above)
Comment (kayibal) [reply sparsity/io.py]: Hm I think numpy closes it will check if that's true for sure though

Comment (michcio1234) [reply sparsity/io.py]: Numpy documentation doesn't say that they close the file.

File: sparsity/io.py (hunk 1 above)
Comment (michcio1234) [reply sparsity/io.py]: Forgot about this one? :P

**Comment by kayibal** [comment]
@michcio1234 Ok to merge?

**Comment by kayibal** [comment]
Argh closed this accidentally

**Pull request merged**